### ISSUE TITLE
0.036 fails t/checksum.t on AIX-5.3

### ISSUE THREAD

**Issue by Tux** [DCIT/perl-CryptX#19]
```
AIX 5.3.0.0/TL12-05  IBM,9115-505 PowerPC_POWER5/1898(2)
This is perl 5, version 14, subversion 2 (v5.14.2) built for aix-64all

t/checksum.t ........................ 1/24
#   Failed test at t/checksum.t line 32.
#          got: '43beb7e8'
#     expected: 'e8b7be43'

#   Failed test at t/checksum.t line 33.
#          got: 'ef7673b3'
#     expected: 'b37376ef'

#   Failed test at t/checksum.t line 34.
#          got: '7309686d'
#     expected: '6d680973'

#   Failed test at t/checksum.t line 35.
#          got: '1929996397'
#     expected: '1835534707'

#   Failed test at t/checksum.t line 36.
#          got: 'e9156e80'
#     expected: '806e15e9'

#   Failed test at t/checksum.t line 37.
#          got: '3910495872'
#     expected: '2154698217'

#   Failed test at t/checksum.t line 39.
#          got: 'ed1f1124'
#     expected: '24111fed'

#   Failed test at t/checksum.t line 40.
#          got: 'c630a41c'
#     expected: '1ca430c6'

#   Failed test at t/checksum.t line 41.
#          got: 'fb4d434d'
#     expected: '4d434dfb'

#   Failed test at t/checksum.t line 42.
#          got: '58829b9f'
#     expected: '9f9b8258'
# Looks like you failed 10 tests of 24.
t/checksum.t ........................ Dubious, test returned 10 (wstat 2560, 0xa00)
Failed 10/24 subtests
```
endianess?
```
$ perl -V:byteorder
byteorder='87654321';
```

**Comment by karel-m** [comment]
> endianess?

Yes :-(

**Comment by karel-m** [comment]
It seems like `src/ltc/headers/tomcrypt_cfg.h` needs some AIX / POWER specific ifdefs like:
```c
/* detect IBM AIX@POWER */
#if defined(_ARCH_PPC) || defined(_ARCH_PPC64)
  #define ENDIAN_BIG
  #if defined(_ARCH_PPC64)
    #define ENDIAN_64BITWORD
  #else
    #define ENDIAN_32BITWORD
  #endif
#endif
```

**Comment by Tux** [comment]
```
$ perl -V:cpp.*
cpp='cpp';
cpp_stuff='42';
cppccsymbols='_AIX=1 _AIX32=1 _AIX41=1 _AIX43=1';
cppflags='-DDEBUGGING -D_ALL_SOURCE -D_ANSI_C_SOURCE -D_POSIX_SOURCE -DUSE_NATIVE_DLOPEN -I/pro/local/include';
cpplast='';
cppminus='';
cpprun='/usr/bin/cpp';
cppstdin='cppstdin';
cppsymbols='_ALL_SOURCE=1 _ANSI_C_SOURCE=1 _BIG_ENDIAN=1 __BIG_ENDIAN__=1 _IBMR2=1 _LARGE_FILE_API= _LONG_LONG=1 _LP64=1 __LP64__=1 _POSIX_C_SOURCE=200112L _POSIX_SOURCE=1 _POWER=1 _XOPEN_SOURCE=600 _XOPEN_SOURCE_EXTENDED=1 __unix=1 __unix__=1';
```
more like _AIX + _BIG_ENDIAN

**Comment by karel-m** [comment]
OK, and those `_AIX`  + `_BIG_ENDIAN` are compiler's built-ins or do they need some `#include <...>`?

**Comment by Tux** [comment]
I'm pretty sure they're compiler builtins

**Comment by karel-m** [comment]
I do not have an AIX machine handy, so could you please try to add the following to `src/ltc/headers/tomcrypt_cfg.h` (after or before already existing `/* detect HP-UX */` section)

```c
/* detect AIX */
#if defined(_AIX) && defined(_BIG_ENDIAN)
  #define ENDIAN_BIG
  #if defined(__LP64__) || defined(_ARCH_PPC64)
    #define ENDIAN_64BITWORD
  #else
    #define ENDIAN_32BITWORD
  #endif
#endif 
```

Or if you have something similar as `p5p-hpux.procura.nl` with AIX and can give me an access I can try it myself.

**Comment by Tux** [comment]
Use key forwarding and from there
```
$ ssh 10.0.0.13
```
should now work

**Comment by karel-m** [comment]
works, thanks

**Comment by brianus** [comment]
I'm getting this same set of errors when I try to compile 0.037 on Darwin 8.5.0 (on an old PPC machine). Can anyone suggest what defines I should be checking for? 

perl -V:cpp.*
`cppflags='-no-cpp-precomp -g -pipe -fno-common -DPERL_DARWIN -no-cpp-precomp -fno-strict-aliasing -I/usr/local/include';
cpp='cpp';
cpp_stuff='42';
cppccsymbols='';
cpplast='-';
cppminus='-';
cpprun='cc -E';
cppstdin='cc -E';
cppsymbols='_BIG_ENDIAN=1 __BIG_ENDIAN__=1 __GNUC__=3 __GNUC_MINOR__=3 __MACH__=1 __STDC__=1';
`
perl -V:byteorder
`byteorder='4321';`

**Comment by karel-m** [comment]
Could you post the output of `cc -E -dM - </dev/null`? (I am assuming that your cc is GNU C)

**Comment by brianus** [comment]
Sure:

```
#define __DBL_MIN_EXP__ (-1021)
#define __FLT_MIN__ 1.17549435e-38F
#define __CHAR_BIT__ 8
#define __WCHAR_MAX__ 2147483647
#define __DBL_DENORM_MIN__ 4.9406564584124654e-324
#define __FLT_EVAL_METHOD__ 0
#define __DBL_MIN_10_EXP__ (-307)
#define __FINITE_MATH_ONLY__ 0
#define __GNUC_PATCHLEVEL__ 1
#define __SHRT_MAX__ 32767
#define __LDBL_MAX__ 1.79769313486231580793728971405301e+308L
#define __APPLE_CC__ 5247
#define __UINTMAX_TYPE__ long long unsigned int
#define __LDBL_MAX_EXP__ 1024
#define __SCHAR_MAX__ 127
#define __USER_LABEL_PREFIX__ _
#define __STDC_HOSTED__ 1
#define __LDBL_HAS_INFINITY__ 1
#define __DBL_DIG__ 15
#define __FLT_EPSILON__ 1.19209290e-7F
#define __LDBL_MIN__ 2.00416836000897277799610805135016e-292L
#define __ppc__ 1
#define __strong 
#define __APPLE__ 1
#define __DECIMAL_DIG__ 33
#define __LDBL_HAS_QUIET_NAN__ 1
#define __DYNAMIC__ 1
#define __GNUC__ 4
#define __DBL_MAX__ 1.7976931348623157e+308
#define __DBL_HAS_INFINITY__ 1
#define __weak 
#define __DBL_MAX_EXP__ 1024
#define __LONG_LONG_MAX__ 9223372036854775807LL
#define __GXX_ABI_VERSION 1002
#define __FLT_MIN_EXP__ (-125)
#define __DBL_MIN__ 2.2250738585072014e-308
#define __DBL_HAS_QUIET_NAN__ 1
#define __REGISTER_PREFIX__ 
#define __NO_INLINE__ 1
#define _ARCH_PPC 1
#define __FLT_MANT_DIG__ 24
#define __VERSION__ "4.0.1 (Apple Computer, Inc. build 5247)"
#define __BIG_ENDIAN__ 1
#define __SIZE_TYPE__ long unsigned int
#define __FLT_RADIX__ 2
#define __LDBL_EPSILON__ 4.94065645841246544176568792868221e-324L
#define __NATURAL_ALIGNMENT__ 1
#define __FLT_HAS_QUIET_NAN__ 1
#define __FLT_MAX_10_EXP__ 38
#define __LONG_MAX__ 2147483647L
#define __FLT_HAS_INFINITY__ 1
#define _BIG_ENDIAN 1
#define __LDBL_MANT_DIG__ 106
#define __WCHAR_TYPE__ int
#define __FLT_DIG__ 6
#define __INT_MAX__ 2147483647
#define __LONG_DOUBLE_128__ 1
#define __FLT_MAX_EXP__ 128
#define __DBL_MANT_DIG__ 53
#define __WINT_TYPE__ int
#define __LDBL_MIN_EXP__ (-968)
#define __MACH__ 1
#define __LDBL_MAX_10_EXP__ 308
#define __DBL_EPSILON__ 2.2204460492503131e-16
#define __INTMAX_MAX__ 9223372036854775807LL
#define __FLT_DENORM_MIN__ 1.40129846e-45F
#define __PIC__ 1
#define __FLT_MAX__ 3.40282347e+38F
#define __FLT_MIN_10_EXP__ (-37)
#define __INTMAX_TYPE__ long long int
#define __GNUC_MINOR__ 0
#define __DBL_MAX_10_EXP__ 308
#define __LDBL_DENORM_MIN__ 4.94065645841246544176568792868221e-324L
#define __PTRDIFF_TYPE__ int
#define __LDBL_MIN_10_EXP__ (-291)
#define __LDBL_DIG__ 31
#define __POWERPC__ 1
```

**Comment by karel-m** [comment]
@brianus could you please try to install this commit 8546ccc ?

**Comment by brianus** [comment]
`All tests successful.`

Thanks!

**Comment by karel-m** [comment]
released as CryptX-0.039

Thanks for your time.
